allchblk.c

```diff
@@ -964,6 +968,13 @@ GC_INNER void GC_freehblk(struct hblk *hbp)
     /* Coalesce with successor, if possible */
       if(0 != nexthdr && HBLK_IS_FREE(nexthdr) && IS_MAPPED(nexthdr)
          && (signed_word)(hhdr -> hb_sz + nexthdr -> hb_sz) > 0
+#     if defined(__CHERI_PURE_CAPABILITY__)
+         /* Bounds of capability should span entire coalesced memory */
+         /* Bounds being larger than blk-size is OK; bounded by the imprecision */
+         /* of original capability obtained from system memory */
+         && (cheri_base_get(hbp) <= cheri_address_get(next))
+         && (cheri_base_get(hbp) + cheri_length_get(hbp)) >= (cheri_address_get(next) + nexthdr->hb_sz)
```

Comment:
This is good as long as the limit isn't 2^64, since that will be represented as 0. That might be Ok here, but you'll know if it's not.

We really want a `cheri_bounds_cover(vaddr_t addr, size_t size)` helper (in `cheriintrin.h`), then this could all be abstracted as a simple `cheri_bounds_cover(next, nexthdr->hb_sz)`.

I [raised an issue](https://github.com/CTSRD-CHERI/llvm-project/issues/599) to discuss this.

Reply:
@djichthys Any thoughts on this?

Reply:
Yes this would be useful to have. It would make the code less verbose / cleaner. 
Currently, we are only ever using these in the *heap*  area and not the stack area.  So practically, it will not hit 2^64.

Reply:
> This is good as long as the limit isn't 2^64

At first glance I thought "this isn't a problem" but then... this *might* depend on the approximation of the bounds. For example, on all current machines I know of, virtual addresses span a maximum of 48 bits though 56 bits will happen soon-ish (I think). If someone got a virtual address that big and it covered a large range, it might span beyond 48 (or 56) bits -- probably not up to (or beyond) 64 bits but one day, maybe... The saturating behaviour does seem like something we should get right now to me.

Reply:
the `hbp` capabilities are only ever stored from what the OS will give it via `mmap()` for e.g.  So the bounds won't overflow as long as the kernel can guarantee that bounds won't overflow (which is a reasonable assumption???). 

If not, I think an additional check of `(addr + len  >= addr )` for each of the arms in the [condition](https://github.com/capablevms/bdwgc/pull/8/files#diff-e3e973b7ade111f4fc2a428e9efea28988f5547754cb703debf607025405f682R976) should be enough without any new CHERI intrinsics / API being introduced.

Reply:
> So the bounds won't overflow as long as the kernel can guarantee that bounds won't overflow (which is a reasonable assumption???).

I don't know if this is guaranteed (or could be guaranteed) as it depends on the particular CHERI platform representation. @jacobbramley might have thoughts.

Reply:
> > So the bounds won't overflow as long as the kernel can guarantee that bounds won't overflow (which is a reasonable assumption???).
> 
> I don't know if this is guaranteed (or could be guaranteed) as it depends on the particular CHERI platform representation. @jacobbramley might have thoughts.

For Morello specifically, it's a reasonable assumption (for `mmap`), but I wouldn't go so far as to guarantee it (especially as I'm not that familiar with BSD and it may differ from Linux in this regard). For CHERI in general, I'm not so sure.

I think I'd be uncomfortable about this in production code, at least without some additional check. Real production code gets away with a lot more, though.

Reply:
I do agree that, in general, the saturating/wrapper/whatever behaviour is worth thinking about early, but not necessarily in a way that holds up @djichthys's patch.

Reply:
Point taken, especially as I think the Cambridge folk will be reluctant to add this behaviour in, meaning we'd have to define it locally.